### ISSUE TITLE
twisted.internet.test.test_sigchld way too aggressive in assuming no previous fd or SIGCHLD handler was installed

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6161]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#6161|
|Type|defect|
|Created|2012-10-25 16:39:57Z|
|Branch|https://github.com/twisted/twisted/tree/sigchld-tests-6161|

`twisted.internet.test.test_sigchld.SetWakeupSIGCHLDTests.setUp` assumes no one has previously set a SIGCHLD handler or registered a wakeup fd. If someone has, that is considered an error. However, any random test that uses the reactor will cause this registration; AFAICT this never caused problems before due to a specific sort order of running tests.

On Ubuntu 12.10:

```
$ trial -z 1 twisted.test.test_udp twisted.internet.test.test_sigchld

...

[ERROR]
Traceback (most recent call last):
  File "/home/itamarst/Devel/Twisted/trunk/twisted/internet/test/test_sigchld.py", line 58, in setUp
    raise RuntimeError("You used some signal APIs wrong!  Try again.")
exceptions.RuntimeError: You used some signal APIs wrong!  Try again.

twisted.internet.test.test_sigchld.SetWakeupSIGCHLDTests.test_uninstallHandler
```

<details><summary>Searchable metadata</summary>

```
trac-id__6161 6161
type__defect defect
reporter__itamar itamar
priority__normal normal
milestone__Python_3_3_Minimal Python 3.3 Minimal
branch__branches_sigchld_tests_6161 branches/sigchld-tests-6161
branch_author__itamarst itamarst
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1351183197000000 1351183197000000
changetime__1351695023000000 1351695023000000
version__None None
owner__itamar itamar

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_1"></a>|@itamarst commented|
|-|-|

(In [36265]) Branching to 'sigchld-tests-6161'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_2"></a>|@itamarst removed owner|
|-|-|

I fixed this by just not raising a `RuntimeError` if a previous handler was installed; `tearDown` already has code to more or less restore the previous signal handler. I also changed the signal handler registered for SIGCHLD to have a name, so that the log message in this situation is more meaningful (the name of the function is logged).

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/sigchld-tests-6161 started.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_3"></a>|@itamarst set owner to @exarkun|
|-|-|

This is, sadly, a dependency for the gireactor port ([#6055](https://github.com/twisted/twisted/issues/6055)).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @itamarst|
|-|-|

> AFAICT this never caused problems before due to a specific sort order of running tests.

Actually it has shown up on buildbot a few times - we have random test order builders.  So it'll be nice to have it fixed. :)

  1. Maybe we don't need the warning either?  It does seem perfectly true that a prior test may have legitimately installed some kind of signal handler.  I don't remember why this check was added in the first place - perhaps only as a way to track down bugs in `test_sigchld.py`.
  1. I'll bike shed on the name of `doNothing` a little bit.  Perhaps `noopSignalHandler`?

That's all, please address these to your satisfaction and merge.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst commented|
|-|-|

  1. It seems kinda untidy to leave global state changes around as a result of tests. If these tests can restore it to where it was when they started (module siginterrupt), why not other tests? So keeping this as reminder seems OK to me.
  1. Done.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_6"></a>|@itamarst set status to `closed`|
|-|-|

(In [36329]) Merge sigchld-tests-6161: Don't fail test_sigchld if SIGCHLD was previously installed.

Author: itamar
Review: exarkun
Fixes: [#6161](https://github.com/twisted/twisted/issues/6161)

The twisted.internet.test.test_sigchld tests were previously somewhat fragile, erroneously raising RuntimeError if a previous test had left a SIGCHLD handler lying around.
